### PR TITLE
Permission less reserve

### PR DESCRIPTION
**Pull request by ilanDoron**
@yaronvel 
this is not for pull. 

only so u can view code.

if u could do some short look...

### CODE REVIEW

File: contracts/PermissionLessReserve.sol [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+    uint32 public numOrdersToAllocate = 60;
+
+    // KNC stakes
+    struct KncStakes {
+        uint128 freeKnc;    // knc that can be used to validate funds
+        uint128 kncOnStake; // per order some knc will move to be kncOnStake. part of it will be used for burning.
+    }
+    
+    //funds data
+    mapping(address => mapping(address => uint)) public makerFunds; // deposited maker funds,
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: what is it? eth or token?

File: contracts/PermissionLessReserve.sol [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+        require(token != address(0));
+        require(_admin != address(0));
+        require(burner != address(0));
+
+        feeBurnerContract = burner;
+        kncToken = knc;
+        reserveToken = token;
+        admin = _admin;
+
+        kncToken.approve(feeBurnerContract, (2**255));
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: please check return value!!!!

File: contracts/PermissionLessReserve.sol [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+    function getConversionRate(ERC20 src, ERC20 dest, uint totalSrcAmount, uint blockNumber) public view returns(uint) {
+
+        require((src == ETH_TOKEN_ADDRESS) || (dest == ETH_TOKEN_ADDRESS));
+        require((src == reserveToken) || (dest == reserveToken));
+        blockNumber; // in this reserve no order expiry == no use for blockNumber. here to avoid compiler warning.
+
+        uint32 orderId;
+
+        if (src == ETH_TOKEN_ADDRESS) {
+            if (isNextOrderTail(BUY_HEAD_ID)) return 0;
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: surely this is not needed. the while loop should take care of it.

File: contracts/PermissionLessReserve.sol [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+                totalDstAmount += order.dstAmount * remainingSrcAmount / order.srcAmount;
+                remainingSrcAmount = 0;
+                break;
+            }
+        }
+
+        if ((remainingSrcAmount != 0) || (totalDstAmount == 0)) return 0; //not enough tokens to exchange.
+
+        //check overflow
+        if (uint(totalDstAmount) * PRECISION < uint(totalDstAmount)) return 0;
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: you probably want to check that total dest amount < max qty.

Comment (yaronvel) [reply contracts/PermissionLessReserve.sol]: beside, something else is wrong here. is totaldestamount is 128 bits, you will never have overflow.

File: contracts/PermissionLessReserve.sol [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+        }
+
+        uint32 orderId;
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount);
+            orderId = BUY_HEAD_ID;
+        } else {
+            require(srcToken.transferFrom(msg.sender, this, srcAmount));
+            require(msg.value == 0);
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: better to switch order of requires

File: contracts/PermissionLessReserve.sol [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+
+            orderId = getNextOrderId(orderId);
+
+            Order memory order = orders[orderId];
+
+            if (order.srcAmount < remainingSrcAmount) {
+                totalDstAmount += order.dstAmount;
+                remainingSrcAmount -= order.srcAmount;
+            } else {
+                totalDstAmount += order.dstAmount * remainingSrcAmount / order.srcAmount;
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: this can surely generate rounding errors.

File: contracts/PermissionLessReserve.sol [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+        uint128 totalDstAmount = 0;
+
+        while (!isNextOrderTail(orderId)) {
+
+            // start with next since head order is dummy order
+            orderId = getNextOrderId(orderId);
+
+            Order memory order = orders[orderId];
+
+            if (order.srcAmount <= remainingSrcAmount) {
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: probably `<` and not `<=`. this way you can avoid `if (remainingSrcAmount == 0) break;`

File: contracts/PermissionLessReserve.sol [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+
+        return true;
+    }
+
+    event NewMakeOrder(uint32 orderId, address indexed maker, bool isEthToToken, uint128 srcAmount, uint128 dstAmount);
+
+    function addMakeOrder(address maker, bool isEthToToken, uint128 srcAmount, uint128 dstAmount,
+        uint32 hintPrevOrder) public returns(bool)
+    {
+        require(maker == msg.sender);
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: then why do you need `maker` param?

File: contracts/PermissionLessReserve.sol [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+        require(validateOrder(maker, isEthToToken, srcAmount, dstAmount));
+
+        uint32 newID = takeOrderId(maker);
+
+        if (hintPrevOrder != 0) {
+
+            addAfterId(maker, newID, srcAmount, dstAmount, hintPrevOrder, isEthToToken? ETH_TO_TOKEN : TOKEN_TO_ETH);
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: I really hate it when there is no return value to check.

File: contracts/PermissionLessReserve.sol [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+        } else {
+
+            if (isEthToToken) {
+                add(maker, newID, srcAmount, dstAmount, BUY_HEAD_ID, ETH_TO_TOKEN);
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: seems you would only need BUY/SELL or ETH_TO_TOKEN/TOKEN_TO_ETH, why do you need both?

File: contracts/PermissionLessReserve.sol [hunk 11]
```diff
@@ -0,0 +1,596 @@
...
+    }
+
+    event MakerDepositedTokens(address indexed maker, uint amountTwei);
+
+    function makerDepositTokens(address maker, uint amountTwei) public {
+        require(maker != address(0));
+
+        require(reserveToken.transferFrom(msg.sender, this, amountTwei));
+
+        makerFunds[maker][reserveToken] += amountTwei;
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: check MAX_QTY before?

File: contracts/PermissionLessReserve.sol [hunk 12]
```diff
@@ -0,0 +1,596 @@
...
+        }
+
+        NewMakeOrder(newID, maker, isEthToToken, srcAmount, dstAmount);
+
+        return true;
+    }
+
+    event MakerDepositedTokens(address indexed maker, uint amountTwei);
+
+    function makerDepositTokens(address maker, uint amountTwei) public {
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: `Token` not `Tokens`.

File: contracts/PermissionLessReserve.sol [hunk 13]
```diff
@@ -0,0 +1,596 @@
...
+
+        require(reserveToken.transferFrom(msg.sender, this, amountTwei));
+
+        makerFunds[maker][reserveToken] += amountTwei;
+        MakerDepositedTokens(maker, amountTwei);
+    }
+
+    event MakerDepositedEth(address indexed maker, uint amountWei);
+
+    function makerDepositEthers(address maker) public payable {
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: `Ether` not `Ethers`.

File: contracts/PermissionLessReserve.sol [hunk 14]
```diff
@@ -0,0 +1,596 @@
...
+        require(maker != address(0));
+
+        makerFunds[maker][ETH_TOKEN_ADDRESS] += msg.value;
+        MakerDepositedEth(maker, msg.value);
+    }
+
+    event MakerDepositedKnc(address indexed maker, uint amountTwei);
+
+    function makerDepositKnc(address maker, uint128 amountTwei) public payable {
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: why 128 bits? do the conversion yourself.

File: contracts/PermissionLessReserve.sol [hunk 15]
```diff
@@ -0,0 +1,596 @@
...
+
+        require(maker != address(0));
```
Comment (yaronvel) [inline contracts/PermissionLessReserve.sol]: stopped reading here.

File: contracts/KyberNetwork.sol [hunk 16]
```diff
@@ -15,16 +15,27 @@ import "./FeeBurnerInterface.sol";
...
+
+    byte constant public PERMISSION_LESS_USAGE_ALLOWED = byte(0);
+    byte constant public PERMISSION_LESS_USAGE_NOT_ALLOWED = byte(1);
+
+//    enum ReserveType {NONE, PERMISSIONED, PERMISSION_LESS_ORDER_BOOK}
+
     uint public negligibleRateDiff = 10; // basic rate steps will be in 0.01%
     KyberReserveInterface[] public reserves;
-    mapping(address=>bool) public isReserve;
+    mapping(address=>uint) public reserveType;
```
[the remark this answers is not in the document]
Comment (ilanDoron) [reply contracts/KyberNetwork.sol]: I need to distinguish between 3 options.
not a reserve or yes (which was done by previous bool)
and permissioned or not.

I thought better do it in one mapping
if bool will need two mappings.
 u think its better?

File: contracts/PermissionLessReserve.sol (hunk 1 above)
Comment (ilanDoron) [reply contracts/PermissionLessReserve.sol]: both.

File: contracts/PermissionLessReserve.sol [hunk 17]
```diff
@@ -0,0 +1,596 @@
...
+    struct FreeOrders {
+        uint32 firstOrderId;
+        uint32 numOrders; //max is 256
+        uint8 maxOrdersReached;
+        uint256 takenBitmap;
+    }
+
+    mapping(address => FreeOrders) makerOrders; //each maker will have orders that will be reused.
+
+    function PermissionLessReserve(FeeBurner burner, ERC20 knc, ERC20 token, address _admin) public {
```
[the remark this answers is not in the document]
Comment (ilanDoron) [reply contracts/PermissionLessReserve.sol]: what do u mean by reserve ID?

**Comment by ilanDoron** [comment]
all fixed.
opened new branch...

**Pull request closed without merging**